challenge_5/go/natdm/diff.go

```diff
@@ -0,0 +1,46 @@
+package natdm
+
+import "math/rand"
+
+// shuffleAndAdd shuffles and adds one character at random
+func shuffleAndAdd(s string) (b byte, out []byte) {
+	// Get a new letter
+	alp := []byte("abcdefghijklmnopqrstuvwxyz")
+	b = alp[rand.Intn(len(alp))]
+
+	// Add it to a new slice that contains the base string/bytes
+	orig := append([]byte(s), b)
+	// Establish return slice
+	out = make([]byte, len(orig))
+
+	// Keep track of indexes already used to pluck a random byte from
+	pickedIndexes := []int{}
+
+	for i := 0; i < len(orig); i++ {
+	draw:
+		r := rand.Intn(len(orig))
+		for _, v := range pickedIndexes {
+			if v == r {
+				goto draw
+			}
+		}
+		pickedIndexes = append(pickedIndexes, r)
+		out[i] = orig[r]
+	}
+	return
+}
+
+func diff(a, b string) (s string) {
+	for _, x := range []byte(b) {
+		for i, y := range []byte(a) {
```

Comment:
Fixed the logic to make a map of chars (byte form) and count for each char in the original. It then loops the new word and each time it finds a character in the map with length > 0, it decrements. If the character is not there or the count is 0, it is an extra and will return it.